### PR TITLE
Return Infinity if simulation terminates early

### PR DESCRIPTION
**Pull request by NicolaCourtier**
A resolution to issue #98.

### CODE REVIEW

File: pybop/costs/error_costs.py [hunk 1]
```diff
@@ -38,7 +38,14 @@ def __call__(self, x, grad=None):
         Computes the cost.
         """
         try:
-            return np.sqrt(np.mean((self.problem.evaluate(x) - self._target) ** 2))
+            prediction = self.problem.evaluate(x)
+
+            if len(prediction) < len(self._target):
+                cost = np.Infinity  # simulation stopped early
```
Comment (BradyPlanden) [inline pybop/costs/error_costs.py:44]: np.inf is probably a better constant to use: https://numpy.org/doc/stable/reference/constants.html

File: pybop/costs/error_costs.py [hunk 2]
```diff
@@ -38,7 +38,14 @@ def __call__(self, x, grad=None):
         Computes the cost.
         """
         try:
-            return np.sqrt(np.mean((self.problem.evaluate(x) - self._target) ** 2))
+            prediction = self.problem.evaluate(x)
+
+            if len(prediction) < len(self._target):
```
Comment (BradyPlanden) [inline pybop/costs/error_costs.py:43]: We can return within the if statement:
```suggestion
            if len(prediction) < len(self._target):
                return np.Infinity  # simulation stopped early
            else:
                return p.sqrt(np.mean((prediction - self._target) ** 2))
```

File: pybop/costs/error_costs.py [hunk 3]
```diff
@@ -57,10 +64,17 @@ def __call__(self, x, grad=None):
         Computes the cost.
         """
         try:
-            return np.sum(
-                (np.sum(((self.problem.evaluate(x) - self._target) ** 2), axis=0)),
-                axis=0,
-            )
+            prediction = self.problem.evaluate(x)
+
+            if len(prediction) < len(self._target):
```
Comment (BradyPlanden) [inline pybop/costs/error_costs.py:69]: Same as above

**Review by BradyPlanden — COMMENTED** [review]
Nice one @NicolaCourtier! I've added a few comments.

Have you thought of any edge-cases on this? In certain design optimisation scenarios, it might be benefitical to return a smaller array (fast-charging comes to mind). There are other issues to solve before this would become the hold-up, but it's probably worth thinking of.

File: pybop/costs/error_costs.py (hunk 1 above)
Comment (NicolaCourtier) [reply pybop/costs/error_costs.py:44]: Thanks for the link!

**Comment by BradyPlanden** [comment]
This looks really good @NicolaCourtier. I think it actually solves #102 as well! I'd like to add a couple things on #102 to help the CMAES assertion, but they don't appear to be as needed. 

For testing this functionality, it would be good to add tests that push the optimiser outside of a stable solution. For reference, I've found setting negative active material fraction below ~0.4 in the SPM/e default parameter set can reliably crash the solver. See here:https://github.com/pybop-team/PyBOP/blob/2c7e74495425ba801aeb2b09670b55fa405465f3/tests/unit/test_parameterisations.py#L157

Happy for this to be merged once the coverage is passing 👍

**Comment by NicolaCourtier** [comment]
I agree with your point above- there will be many edge cases where this is not the appropriate fix, but for the two Costs that we have now (RootMeanSquaredError and SumSquaredError) where there is a target to match, I think it gives predictable behaviour.

I think the downside would be that any dataset which spends time at eiher voltage limit could be very hard for the optimiser to find, because parameter sets that marginally break the limits will return an infinite cost even though they could be quite close to the optimal values. The first suggestion (extending the output array - there would be several choices we could make here) would give the optimiser more information in such scenarios and therefore possibly better performance in such edge cases.

File: pybop/costs/error_costs.py [hunk 4]
```diff
@@ -38,7 +38,12 @@ def __call__(self, x, grad=None):
         Computes the cost.
         """
         try:
-            return np.sqrt(np.mean((self.problem.evaluate(x) - self._target) ** 2))
+            prediction = self.problem.evaluate(x)
+
+            if len(prediction) < len(self._target):
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/costs/error_costs.py:43]: I think we should leave this for now, we can expect to receive shorter solutions but I can't think of a reason for longer at the moment. I don't think that this fix should catch longer solutions as they would have an alternative explanation.

File: pybop/costs/error_costs.py [hunk 5]
```diff
@@ -47,20 +52,32 @@ def __call__(self, x, grad=None):
...
         Computes the cost.
         """
         try:
-            return np.sum(
-                (np.sum(((self.problem.evaluate(x) - self._target) ** 2), axis=0)),
-                axis=0,
-            )
+            prediction = self.problem.evaluate(x)
+
+            if len(prediction) < len(self._target):
```
[the remark this answers is not in the document]
Comment (NicolaCourtier) [reply pybop/costs/error_costs.py:73]: Same as above

**Review by BradyPlanden — APPROVED** [review]
